From a0943bd277ed9b5ecaf1c7c1ace7590b1b6da37d Mon Sep 17 00:00:00 2001
From: Shilin Victor <chrono.monochrome@gmail.com>
Date: Mon, 20 Aug 2018 22:57:43 +0300
Subject: [PATCH] A memory leak occurs. RS::threadProc()

Bug Details:
When realloc() fails with RS::threadProc(), the memory secured before realloc() has not been released and a memory leak occurs.

Modification:
When memory could not be allocated, free() processing was added.

Supplement:
Detected with the static analysis tool.

diff --git a/cpp/RenderScript.cpp b/cpp/RenderScript.cpp
old mode 100644
new mode 100755
index 92e456f..2246c08
--- a/cpp/RenderScript.cpp
+++ b/cpp/RenderScript.cpp
@@ -219,7 +219,13 @@
 
         if (receiveLen >= rbuf_size) {
             rbuf_size = receiveLen + 32;
-            rbuf = realloc(rbuf, rbuf_size);
+            void *tmpBuf = realloc(rbuf, rbuf_size);
+            if (tmpBuf) {
+                rbuf = tmpBuf;
+            } else {
+                free(rbuf);
+                rbuf = NULL;
+            }
         }
         if (!rbuf) {
             ALOGE("RS::message handler realloc error %zu", rbuf_size);
